test/functional/p2p_segwit.py

```diff
@@ -252,11 +252,11 @@ def test_unnecessary_witness_before_segwit_activation(self):
         tx3.vin.append(CTxIn(COutPoint(tx2.sha256, 0), CScript([p2sh_program])))
         tx3.vout.append(CTxOut(tx2.vout[0].nValue-1000, scriptPubKey))
         tx3.wit.vtxinwit.append(CTxInWitness())
-        tx3.wit.vtxinwit[0].scriptWitness.stack = [b'a'*400000]
+        tx3.wit.vtxinwit[0].scriptWitness.stack = [b'a' * 400]
         tx3.rehash()
-        # Note that this should be rejected for the premature witness reason,
-        # rather than a policy check, since segwit hasn't activated yet.
-        test_transaction_acceptance(self.nodes[1].rpc, self.std_node, tx3, True, False, b'no-witness-yet')
+        # This will be rejected due to a policy check:
+        # No witness is allowed, since it is not a witness program but a p2sh program
```

Comment:
Just remove this entire block (everything from `We'll add an unnecessary witness to this transaction that would cause` down). `bad-witness-nonstandard` is tested further down.

Reply:
Please note that `bad-witness-nonstandard` is a general reject reason and as far as I can see none of the tests below check for unnecessary witness. So I guess it is fine to keep this as is.

Reply:
I disagree. The entire `test_unnecessary_witness_before_segwit_activation()` subtest should be removed, since this PR updates the logic so there's no difference in behaviour before and after activation. Leaving this test as it is is misleading, since it implies that there's some difference in logic before and after activation.

Reply:
Please note that the policy difference only has effect on transactions (and not blocks). This test is testing that blocks can not include unnecessary witness. It is also testing that unnecessary witness that is added to transactions "in-flight" (e.g. by malicious peers) does not add them to the rejection cache and that they'd be accepted just fine when sent without the witness.

Reply:
I have amended the misleading comment in my last commit.

Reply:
This subtest should be in `test_non_standard_witness()`, not in `test_unnecessary_witness_before_segwit_activation()`

Reply:
I have moved it to a separate test case so that the symbols don't collide